packages/language-server/test/completion-items-spec.ts

```diff
@@ -25,28 +29,245 @@ describe("the UI5 language assistant Code Completion Services", () => {
+      `${description}: ${filterText} does not contain all characters from ${text} in order`
+    ).to.be.true;
+  }
+
+  /** Check the suggestions will be displayed to the user according to the range and filter text */
+  function assertSuggestionsAreValid(
+    suggestions: CompletionItem[],
+    xmlSnippet: string
+  ): void {
+    const xmlText = xmlSnippet.replace("⇶", "");
+    const offset = xmlSnippet.indexOf("⇶");
+    const doc: TextDocument = createTextDocument("xml", xmlText);
+    const position = doc.positionAt(offset);
+
+    forEach(suggestions, suggestion => {
+      expectExists(suggestion.textEdit, "suggestion contains a textEdit");
+      assertRangeContains(
+        suggestion.textEdit.range,
+        position,
+        suggestion.label
+      );
+      // The filter text is checked until the position in the document
+      // (for example, we can replace "Ab⇶cd" with "Abzzz" even though "c" and "d" aren't in "Abzzz")
+      const checkedRange = {
+        start: suggestion.textEdit?.range.start,
+        end: position
+      };
+      assertFilterMatches(
+        suggestion.filterText ?? suggestion.label,
+        doc.getText(checkedRange),
+        suggestion.label
+      );
+    });
+  }
+
+  function getTextInRange(
+    xmlSnippet: string,
+    range: Range | undefined
+  ): string {
+    const xmlText = xmlSnippet.replace("⇶", "");
+    const doc: TextDocument = createTextDocument("xml", xmlText);
+    return doc.getText(range);
+  }
+
   it("will get completion values for UI5 class", () => {
     const xmlSnippet = `<GridLi⇶`;
     const suggestions = getSuggestions(xmlSnippet);
     const suggestionNames = map(suggestions, suggestion => ({
       label: suggestion.label,
-      text: getTagName(suggestion.insertText)
+      tagName: getTagName(suggestion.textEdit),
+      replacedText: getTextInRange(xmlSnippet, suggestion.textEdit?.range)
+    }));
+    const suggestionKinds = uniq(
+      map(suggestions, suggestion => suggestion.kind)
+    );
+
+    expect(suggestionNames).to.deep.equalInAnyOrder([
+      { label: "GridList", tagName: "f:GridList", replacedText: "GridLi" },
+      {
+        label: "GridListItem",
+        tagName: "f:GridListItem",
+        replacedText: "GridLi"
+      }
+    ]);
+
+    expect(suggestionKinds).to.deep.equal([CompletionItemKind.Class]);
+  });
+
+  it("will get completion values for UI5 class by fully qualified name", () => {
+    const xmlSnippet = `<sap.m.Busy⇶`;
+    const suggestions = getSuggestions(xmlSnippet);
+    const suggestionNames = map(suggestions, suggestion => ({
+      label: suggestion.label,
+      tagName: getTagName(suggestion.textEdit),
+      replacedText: getTextInRange(xmlSnippet, suggestion.textEdit?.range)
+    }));
+    const suggestionKinds = uniq(
+      map(suggestions, suggestion => suggestion.kind)
+    );
+
+    expect(suggestionNames).to.deep.equalInAnyOrder([
+      {
+        label: "BusyDialog",
+        tagName: "m:BusyDialog",
```

Comment:
why do we assert on the `tagName`? could/should we assert directly on the textEdit to help
readers of the code understand textEdit scenario from literal examples?

Reply:
We can assert on the entire text edit (its `newText` property) but that includes things like closing brackets and tab stops. I think maybe we should have only a few tests that check these things specifically and most tests can check only the tag name (so that if we change the tab stops for example we won't need to change a lot of tests).

Reply:
can we represent the `textEdit` in some simple string literal form?
The reasoning is to help understand the flow from a technical perspective.